feature: artifact-helper

pvbouwel opened the pull request:
Allow uploading artifacts using:
```
from openeo.extra.artifacts import ArtifactHelper

artifact_helper = ArtifactHelper.from_openeo_connection(connection)
storage_uri = artifact_helper.upload_file(object_name, src_file_path)
presigned_uri = artifact_helper.get_presigned_url(storage_uri)
```

Where object_name is a logical string name that will be used to give a unique name in the backend storage.
And src_file_path points to a local file that needs to be stored remotely.

This MVP version still has the sts and s3 URIs hard coded.  This is temporary. It is possible to expand the links that are exposed at the root_url (https://openeo.dataspace.copernicus.eu/openeo/1.2/)  to include rels like 'artifacts-s3-endpoint' and 'artifacts-sts-endpoint'. That way different artifact backends can be chosen based on the OpenEO environment.

soxofaan reviewed (commented):
this is quite a large diff :smile:  and I haven't gone deep yet, but here are some initial notes

soxofaan commented on openeo/extra/artifacts/internal_s3/artifact_helper.py:
```diff
@@ -0,0 +1,68 @@
+    def _get_upload_key(self, object_name: str) -> str:
+        return f"{self._get_upload_prefix()}{object_name}"
+    
+    def upload_file(self, object_name: str, src_file_path: str | Path) -> S3URI:
```
Isn't it more logical to put `src_file_path` argument before `object_name` (like bash `cp src dst`).
It would also allow to make `object_name` optional  provide a autogenerated name.

minor note: `src_file_path` looks a bit long as name, could be just `path`, `source` or `src` I think

pvbouwel replied:
Yes makes a lot of sense as object name naturally can be the filename by default and only special cases would want to chose the target object name.

Incorporated feedback including minor note.

pvbouwel commented on openeo/extra/artifacts/internal_s3/artifact_helper.py:
```diff
@@ -0,0 +1,68 @@
+from openeo.extra.artifacts.internal_s3.config import S3Config
+from openeo.extra.artifacts.internal_s3.model import S3URI
+
+if TYPE_CHECKING:
```
Will fix cleanup because this guard indeed does not solve any problem.

pvbouwel commented on openeo/extra/artifacts/internal_s3/config.py:
```diff
@@ -0,0 +1,112 @@
+from packaging.version import Version
+
+
+if Version(botocore.__version__) < Version("1.36.0"):
```
Since 1.36.0 is only from 2025 I prefer to not force it onto users but it makes sense to not take another dependency so will refactor to use ComparableVersion

pvbouwel commented on openeo/extra/artifacts/internal_s3/config.py:
```diff
@@ -0,0 +1,112 @@
+        Hard coding since connection does not allow automatic determining config yet.
+        """
+        if self.s3_endpoint_url is None:
+            self.s3_endpoint_url = "https://s3.waw3-1.openeo.v1.dataspace.copernicus.eu"
```
I agree the hardcoding is not preferred and it is meant as a temporary stop gap.
I don't think there is a proper default value hence the None of the dataclass seemed sensible.

The reason for this boiler plate style code is because it can then be easily replaced with a request to the openeo_URL to check metadata of the backend and to configure based on that.

pvbouwel replied:
Config is now fetched using the connection url.

pvbouwel commented on openeo/extra/artifacts/internal_s3/model.py:
```diff
@@ -0,0 +1,56 @@
+class AWSSTSCredentials:
+    AWS_ACCESS_KEY_ID: str
+    AWS_SECRET_ACCESS_KEY: str
+    AWS_SESSION_TOKEN: str
```
It is left over but from functionality that is not there so there is no reason to not follow Python styling. I'll refactor.

pvbouwel commented on openeo/extra/artifacts/internal_s3/model.py:
```diff
@@ -0,0 +1,56 @@
+    def from_str(cls, uri: str) -> S3URI:
+        s3_prefix = "s3://"
+        if uri.startswith(s3_prefix):
+            without_prefix = uri[len(s3_prefix):]
```
refactored to use urlparse

pvbouwel commented on openeo/extra/artifacts/__init__.py:
```diff
@@ -0,0 +1,9 @@
+# Hard coded at this time but this could be a builder that depending on connection builds a client for the Storage
+from openeo.extra.artifacts.internal_s3.artifact_helper import S3ArtifactHelper as ArtifactHelper
```
Well I wanted users to use ArtifactHelper rather than S3ArtifactHelper as it is meant as a generic builder that can be extended when needed. I was not aware of how it would get processed by the sphinx doc system.

Would it be better to already have the builder eventhough it has a dummy implementation for now? 
https://github.com/Open-EO/openeo-python-client/pull/750/commits/a3eea6214d998dff461f39e6ec12c1fe0fb79969

pvbouwel commented on openeo/extra/artifacts/__init__.py:
```diff
@@ -0,0 +1,9 @@
+artifact_helper = ArtifactHelper.from_openeo_connection(connection)
+storage_uri = artifact_helper.upload_file(object_name, src_file_path)
+presigned_uri = artifact_helper.get_presigned_url(storage_uri)
+"""
```
It was documentation but moved it now with addressing other comment

pvbouwel commented on openeo/extra/artifacts/config.py:
```diff
@@ -0,0 +1,36 @@
+        """
+        if not hasattr(self, _METADATA_LOADED):
+            return False
+        return getattr(self, _METADATA_LOADED)
```
This avoids the need of initialization which makes it nicely extendable by dataclasses.

pvbouwel commented on openeo/extra/artifacts/artifact_helper.py:
```diff
@@ -0,0 +1,54 @@
+        These URIs should be kept secret as they provide access to the data.
+        """
+
+    def __init__(self, config: StorageConfig):
```
The idea was to have important methods that users are aware of first. Users should never call this init so hiding it was a concious decission. If you still prefer it to be on top I am open to move it but personally I like it being hidden.

pvbouwel commented on openeo/extra/artifacts/internal_s3/__init__.py:
I wasn't aware same convention was used for modules

pvbouwel commented on openeo/extra/artifacts/internal_s3/config.py:
```diff
@@ -0,0 +1,112 @@
+
+    @staticmethod
+    def _remove_protocol_from_uri(uri: str):
+        uri_separator = "://"
```
For the other comment I moved to urlparse as I indeed needed real manipulation of the URI. For this one however I just need to strip the protocol. I don't see how full parsing of the URL will help to simplify. The problem with parsing is that it is easy to miss edge cases and have query parameters dropped or other special behavior due to certain characters. By just keeping the string opaque no special handling is needed.

pvbouwel commented on openeo/extra/artifacts/internal_s3/sts.py:
```diff
@@ -0,0 +1,37 @@
+        auth = conn.auth
+        assert auth is not None
+        if not isinstance(auth, BearerAuth):
+            raise ValueError("Only connections that use federation are allowed.")
```
I'll rephrase it.

pvbouwel commented on openeo/extra/artifacts/internal_s3/sts.py:
```diff
@@ -0,0 +1,37 @@
+            sts.assume_role_with_web_identity(
+                RoleArn=self._get_aws_access_role(),
+                RoleSessionName=auth_token[1],
+                WebIdentityToken=auth_token[2],
```
Support for others can be added when there is a need currently JWT is the only one we support but in the end the only important part is that a token is verifiable .

The RoleSessionName is not really that important to the authentication process but it would make sense to use a more useful value. I will adapt to  f"artifact-helper-{datetime.datetime.now(datetime.UTC).strftime('%Y%m%d%H%M%S')}"

pvbouwel commented on openeo/extra/artifacts/uri.py:
```diff
@@ -0,0 +1,15 @@
+        """factory method to create a typed object from its string representation"""
+
+    @abstractmethod
+    def __str__(self):
```
Make sense went for `to_string`

pvbouwel commented:
@soxofaan provided docs as requested. I did move around code but mostly renames of files to start with _ to discourage users from entering.

Other changes:

- introduced feature_flags for s3sts such that backend can advertise what is supports
- now decided on advertised feature flag whether to enable/disable checksumming
- more tests
- Added JSON lexer for sphinx as JSON would not render properly otherwise

soxofaan commented on openeo/extra/artifacts/_s3sts/artifact_helper.py:
```diff
@@ -0,0 +1,105 @@
+        self._s3: Optional[S3Client] = None
+
+    @classmethod
+    def _from_openeo_connection(cls, conn: Connection, config: S3STSConfig) -> S3STSArtifactHelper:
```
this applies to other places in the PR too I think

soxofaan commented on docs/_ext/jsonlexer.py:
```diff
@@ -0,0 +1,5 @@
+
+
+def setup(app):
+    app.add_lexer("json", JsonLexer)
```
Inspired by this bit of the PR I just tried with 98fae3d62f4c8fa8e51eb3ffe17ebf47990f9e87  to use `.. code-block:: json`  in docs/udp.rst  and that seems to work just fine as far as I see on https://open-eo.github.io/openeo-python-client/udp.html

soxofaan commented on docs/api-artifacts.rst:
```diff
@@ -0,0 +1,185 @@
+               "sts_endpoint": "https://my.sts.test"              // Where STS API calls are sent
+             },
+             "id": "s3",                                          // This id is a logical name
+             "type": "S3STSConfig"                                // The config type of the ArtifactHelper
```
put id and type above config?

soxofaan commented on openeo/extra/artifacts/_backend.py:
```diff
@@ -0,0 +1,114 @@
+        except IndexError:
+            return ProviderConfig("id", "type", {}, exc=NoAdvertisedProviders())
+        except ArtifactsException as e:
+            return ProviderConfig("id", "type", {}, exc=e)
```
is it intentional to use id "id" and type "type" here? Or is that still placeholder code?

if there is no known id/type, I'd find it more natural to use `None` or a string like "n/a" or "unknown"

soxofaan commented on openeo/extra/artifacts/_artifact_helper_abc.py:
```diff
@@ -0,0 +1,95 @@
+    This class defines the *interface* that an artifact helper should implement and support. This is used by OpenEO users
+    willing to manage artifacts.
+
+    Instances that implement it get created by the `openeo.extra.artifacts.artifact_helper.ArtifactHelper` factory
```
`ArtifactHelper` is not a thing anymore, right?

soxofaan commented on openeo/extra/artifacts/exceptions.py:
```diff
@@ -0,0 +1,61 @@
+        self.type_id = type_id
+
+    def __str__(self):
+        return f"The OpenEO backend does not support {self.type_id}"
```
I'm not sure about the pattern to override `__str__` in exception subclasses. Often, exceptions are `repr`ed instead of `str`ed, and then you won't get that error message.

I think it's more robust and simpler, to do everything in `__init__`, e.g.

```python
class UnsupportedArtifactsType(ArtifactsException):
    def __init__(self, type_id: str):
        super().__init__(f"The OpenEO backend does not support {type_id}")
```

(if you would still want to be able to recover the `type_id` from the exception, you can still attach it as attribute in `__init__` of course, but I left that out for clarity)

soxofaan commented on docs/api-artifacts.rst:
```diff
@@ -0,0 +1,185 @@
+How does it work ?
+==================
+
+1) :py:meth:`openeo.extra.artifacts.artifact_helper.ArtifactHelper.from_openeo_connection` is a factory method that
```
yeah when doing incremental builds during development, sphinx sometimes misses dangling pointers

soxofaan commented on docs/api-artifacts.rst:
```diff
@@ -0,0 +1,185 @@
+
+1. Connect to your openEO backend
+2. Create an artifact helper from your openEO connection
+3. Upload your file using the artifact helper and optionally get a presigned URI
```
it might be important to settle some basics of the spec, e.g. if "artifacts" as toplevel key in the capabilities doc isn't too generic as a name

soxofaan commented:
I would also add a changelog entry along the lines of

```
- Initial/experimental implementation of artifact upload helper ([Open-EO/openeo-api#566](https://github.com/Open-EO/openeo-api/issues/566))
```

soxofaan commented on docs/api-artifacts.rst:
```diff
@@ -0,0 +1,200 @@
+
+.. warning::
+    This is a new experimental API, subject to change.
+
```
also add this note:

```

.. versionadded:: 0.44.0

```

(assuming that will be the release it will be added)

pvbouwel replied:
Squashed current state and added the version + changelog as a separate commit

soxofaan commented on docs/api-artifacts.rst:
```diff
@@ -0,0 +1,116 @@
+====================================
+API: openeo.extra.artifacts
+====================================
+.. versionadded:: 0.44.0
```
```suggestion
.. versionadded:: 0.45.0
```

(0.44 was already released in mean time)

soxofaan reviewed (commented):
I'd merge latest master in this feature branch
